api/sip/client.py

```diff
@@ -252,127 +248,94 @@ def __init__(self, target_server, target_port, login_user_id=None,
             escaped = self.separator
         self.separator_re = re.compile(escaped + "([A-Z][A-Z])")
 
+        self.sequence_number = 0
```

Comment:
The sequence number needs to be associated with the session. Previously the session and the SIPClient class were one and the same. Now there's no separate "session" object. If a SIPClient is used by multiple threads, we'll have multiple sessions going on at once and sharing a sequence number.

Try creating a SIPSession object which manages the connection and the sequence number.

Reply:
How about if the connection just moves back into the class, so the session and object are one in the same? Does that address your concern?

Reply:
@leonardr I see the elegance in the solution you propose of creating a `SIPSession` object and I will look into making that change. 

I am wondering about your comment on threading though. Is this a theoretical concern in case this is used by multiple threads in the future or do you see a threading problem as the code is used right now? 

My understanding is that currently each thread gets its own `SIP2AuthenticationProvider` and each of those has its own `SIPClient` so I don't see where `SIPClient` can be called into from multiple thread currently. Just want to make sure I am understanding correctly.

Reply:
The multi-threading thing is theoretical for us, but at the moment this is the most functional Python SIP2 client on the Internet, so I'd like to keep it usable in someone else's multi-threaded scenario (or our future scenario).

As I said in another comment, it should also be fine to treat `SIPClient` as lasting the length of a SIP session. In that case you would establish a connection immediately and the sequence number could be associated with the `SIPClient` object.  `SIPClient`s would be disposable, `SIPAuthenticationProvider` would create them all the time instead of holding on to them, and you wouldn't use them across threads.